client/network-gossip/src/bridge.rs

```diff
@@ -0,0 +1,299 @@
+	/// the message's topic. No validation is performed on the message, if the
+	/// message is already expired it should be dropped on the next garbage
+	/// collection.
+	pub fn register_gossip_message(
+		&self,
+		topic: B::Hash,
+		message: Vec<u8>,
+	) {
+		let message = ConsensusMessage {
+			engine_id: self.engine_id,
+			data: message,
+		};
+
+		self.inner.lock().state_machine.register_message(topic, message);
+	}
+
+	/// Broadcast all messages with given topic.
+	pub fn broadcast_topic(&self, topic: B::Hash, force: bool) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.broadcast_topic(&mut *inner.context, topic, force);
+	}
+
+	/// Get data of valid, incoming messages for a topic (but might have expired meanwhile)
+	pub fn messages_for(&self, topic: B::Hash)
+		-> mpsc::UnboundedReceiver<TopicNotification>
+	{
+		self.inner.lock().state_machine.messages_for(self.engine_id, topic)
+	}
+
+	/// Send all messages with given topic to a peer.
+	pub fn send_topic(
+		&self,
+		who: &PeerId,
+		topic: B::Hash,
+		force: bool
+	) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.send_topic(&mut *inner.context, who, topic, self.engine_id, force)
+	}
+
+	/// Multicast a message to all peers.
+	pub fn gossip_message(
+		&self,
+		topic: B::Hash,
+		message: Vec<u8>,
+		force: bool,
+	) {
+		let message = ConsensusMessage {
+			engine_id: self.engine_id,
+			data: message,
+		};
+
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+		inner.state_machine.multicast(&mut *inner.context, topic, message, force)
+	}
+
+	/// Send addressed message to the given peers. The message is not kept or multicast
+	/// later on.
+	pub fn send_message(&self, who: Vec<network::PeerId>, data: Vec<u8>) {
+		let mut inner = self.inner.lock();
+		let inner = &mut *inner;
+
+		for who in &who {
+			inner.state_machine.send_message(&mut *inner.context, who, ConsensusMessage {
+				engine_id: self.engine_id,
+				data: data.clone(),
+			});
+		}
+	}
+
+	/// Notify everyone we're connected to that we have the given block.
+	///
+	/// Note: this method isn't strictly related to gossiping and should eventually be moved
+	/// somewhere else.
+	pub fn announce(&self, block: B::Hash, associated_data: Vec<u8>) {
+		self.inner.lock().context_ext.announce(block, associated_data);
+	}
+
+	/// Notifies the sync service to try and sync the given block from the given
+	/// peers.
+	///
+	/// If the given vector of peers is empty then the underlying implementation
+	/// should make a best effort to fetch the block from any peers it is
+	/// connected to (NOTE: this assumption will change in the future #3629).
+	///
+	/// Note: this method isn't strictly related to gossiping and should eventually be moved
+	/// somewhere else.
+	pub fn set_sync_fork_request(&self, peers: Vec<network::PeerId>, hash: B::Hash, number: NumberFor<B>) {
```

Comment:
Is this really something that should be provided by the gossip crate? Not a blocker, just seems less than ideal.

Reply:
This is a much deeper refactoring than the scope of this PR.

Reply:
How about adding a note here, creating a GitHub issue and following up on it later? 

(I am happy to help tackle this in a follow up pull request.)

Reply:
Would be fixed via https://github.com/paritytech/substrate/pull/4359.